shared/android/app/src/main/java/io/keybase/ossifrage/KBPushNotifier.java

```diff
@@ -1,56 +1,214 @@
+import android.graphics.PorterDuffXfermode;
+import android.graphics.Rect;
 import android.os.Bundle;
 import android.support.v4.app.NotificationCompat;
+import android.support.v4.app.NotificationCompat.MessagingStyle;
 import android.support.v4.app.NotificationManagerCompat;
+import android.support.v4.app.Person;
+import android.support.v4.graphics.drawable.IconCompat;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
 
-import com.dieam.reactnativepushnotification.modules.RNPushNotificationHelper;
+import keybase.ChatNotification;
+import keybase.Message;
+import keybase.PushNotifier;
 
 public class KBPushNotifier implements PushNotifier {
-    private final Context context;
-    private Bundle bundle;
+  private final Context context;
+  private Bundle bundle;
+
+  private SmallMsgRingBuffer convMsgCache;
+
+  private MessagingStyle buildStyle(Person person) {
+    MessagingStyle style = new MessagingStyle(person);
+    if (convMsgCache != null) {
+      for (MessagingStyle.Message msg : convMsgCache.summary()) {
+        style.addMessage(msg);
+      }
+    }
+
+    return style;
+  }
+
+  public KBPushNotifier(Context ctx, Bundle bundle) {
+    this.context = ctx;
+    this.bundle = bundle;
+  }
+
+  public void setMsgCache(SmallMsgRingBuffer convMsgCache) {
+    this.convMsgCache = convMsgCache;
+  }
+
+  // From: https://stackoverflow.com/questions/11932805/cropping-circular-area-from-bitmap-in-android
+  private static Bitmap getCroppedBitmap(Bitmap bitmap) {
+    Bitmap output = Bitmap.createBitmap(bitmap.getWidth(),
+      bitmap.getHeight(), Bitmap.Config.ARGB_8888);
+    Canvas canvas = new Canvas(output);
+
+    final int color = 0xff424242;
+    final Paint paint = new Paint();
+    final Rect rect = new Rect(0, 0, bitmap.getWidth(), bitmap.getHeight());
+
+    paint.setAntiAlias(true);
+    canvas.drawARGB(0, 0, 0, 0);
+    paint.setColor(color);
+    canvas.drawCircle(bitmap.getWidth() / 2, bitmap.getHeight() / 2,
+      bitmap.getWidth() / 2, paint);
+    paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
+    canvas.drawBitmap(bitmap, rect, rect, paint);
+    return output;
+  }
+
+  // Controls the Intent that gets built
+  private PendingIntent buildPendingIntent(Bundle bundle) {
+    Intent open_activity_intent = new Intent(context, MainActivity.class);
+    open_activity_intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
+    open_activity_intent.setPackage(context.getPackageName());
+    open_activity_intent.putExtra("notification", bundle);
+
+    PendingIntent pending_intent = PendingIntent.getActivity(this.context, 0, open_activity_intent,
+      PendingIntent.FLAG_UPDATE_CURRENT);
+
+    return pending_intent;
+  }
+
+  private IconCompat getKeybaseAvatar(String avatarUri) {
+    IconCompat icon = null;
+
+    if (!avatarUri.isEmpty()) {
```

Comment:
To reduce nesting, I'd prefer:
```suggestion
    if (avatarUri.isEmpty()) return null;
    ...
```
Then you don't even need the `icon` declaration above, instead returning directly where it's needed.